README.md
# Tangent Images for Mitigating Spherical Distortion
![Tangent Images](./images/figure_1.png)

## Dependencies

This repository is designed to be used with PyTorch. In its current form, it has been tested with PyTorch 1.3.1.

#### Python

* PyTorch 1.3.1
* numpy
* matplotlib (plotting examples)
* scikit-image (io)
* plyfile (for saving pointclouds)
* visdom (training visualizations)
* OpenCV 3.4.2 (note that this version is required for proper SIFT detection)

#### Other dependencies

* Eigen (for CGAL--the aptitude package version works)
* CGAL 4.11+ (the aptitude package version works)
* CUDA 10.1 (optional for PyTorch GPU support)


## Installation

#### Installing dependencies
To install the necessary Python dependencies, we recommend creating a Conda environment to install from the provided `tangent-images.yml` file. [Directions can be found here.](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-from-an-environment-yml-file)

CGAL is a required dependency. You can [install the most recent version from source](https://doc.cgal.org/latest/Manual/general_intro.html) (header only is fine). If you use the aptitude package manager on Linux, you can alternatively install the library with the command:

```
sudo apt install libcgal-dev
```

The CGAL dependency relies on Eigen, so you may also need to install:

```
sudo apt install libeigen3-dev
```

#### Installing tangent-images

 1. Make sure you have activated your Conda environment or are otherwise using the desired Python environment
 2. Navigate to `<top-level>/package`
 3. `python setup.py install`


You should be able to test if the installation was successful by running the example scripts in [examples](./examples).

## Experiments

All experiments are included in the [experiments](./experiments) folder. Each experiment subdirectory has a README file explaining how to setup and run each experiments. Where relevant, we have included the pre-trained models corresponding to our published results. Note that both the transfer learning and semantic segmentation experiments are in [`experiments/semantic_segmentation`](experiments/semantic_segmentation).

**TODO**: Semantic segmentation, transfer learning, and SIFT experiments forthcoming.

## Attribution

If you find this repository useful for your own work, please make sure to cite our paper:

```
@article{eder2019tangent,
    title={Tangent Images for Mitigating Spherical Distortion},
    author={Marc Eder and Mykhailo Shvets and John Lim and Jan-Michael Frahm},
    eprint={arXiv:1912.09390},
    year={2019}
}
```

HISTORY.rst
=======
HISTORY
=======

0.0.6 (2019-12-17)
------------------
* Included model for shape classification
* README for semantic segmentation
* Beginning of SIFT experiments

0.0.5 (2019-12-16)
------------------
* Results for ShapeNet
* Add distorted MNIST experiment
* Main README file

0.0.4 (2019-12-16)
------------------
* Updates to image ops utilities (and one tangent image utility)
* Finished SIFT examples

0.0.3 (2019-12-16)
------------------
* Renamed some utility functions and added some others
* Finished tangent image generation examples

0.0.2 (2019-12-14)
------------------
* Shape classification experiment
* Upgrade to PyTorch 1.3.1

0.0.1 (2019-12-12)
------------------
* bumpversion

0.0.0 (2019-12-10)
------------------
* Backend
* Initial examples
* Semantic segmentation code

examples/compare_sift_keypoints.py
import torch
import torch.nn.functional as F
from tangent_images.util import *
from tangent_images.nn import *
import matplotlib.pyplot as plt
from skimage import io, transform

# ----------------------------------------------
# Parameters
# ----------------------------------------------
base_order = 1  # Base sphere resolution
sample_order = 10  # Determines sample resolution (10 = 2048 x 4096)
scale_factor = 1  # How much to scale input equirectangular image by
save_ply = True  # Whether to save the PLY visualizations too

# ----------------------------------------------
# Compute necessary data
# ----------------------------------------------
# Generate the base icosphere
icosphere = generate_icosphere(base_order)

# Number of samples per tangent image dimension
num_samples = compute_num_samples(base_order, sample_order)

# Sampling resolutions
sampling_resolution = get_sampling_resolution(base_order)

# Corners of tangent planes in spherical coordinates (N x 4 x 2)
corners = tangent_image_corners(
    icosphere, num_samples, num_samples, sampling_resolution / num_samples,
    sampling_resolution / num_samples, 'face').squeeze(0)

# Convert the spherical corners to corners in 3D space
corners = convert_spherical_to_3d(corners)

# ----------------------------------------------
# Load and preprocess the image
# ----------------------------------------------
img = io.imread('earthmap4k.jpg').astype(np.float32)
img = (255 * transform.rescale(
    img / 255,
    scale=scale_factor,
    order=1,
    anti_aliasing=True,
    mode='constant',
    multichannel=True)).astype(np.uint8)

# Convert image to torch format
img = torch.from_numpy(img).float().permute(2, 0, 1).contiguous()

# Module to resample the image to UV textures
resample_to_uv_layer = ResampleToUVTexture(img.shape[-2:], base_order,
                                           sample_order, 1)

# Resample the image to N tangent images (out: 3 x N x H x W)
tex_image = resample_to_uv_layer(img.unsqueeze(0)).squeeze(0).byte()

if save_ply:
    # Create and save textured sphere at sample order
    viz_icosphere = generate_icosphere(sample_order)
    unresampler = Unresample('bispherical')
    sample_map = equirectangular_to_sphere_resample_map(img.shape[-2:],
                                                        viz_icosphere)
    rgb_vertices = unresampler(img.unsqueeze(0), sample_map).squeeze()
    viz_icosphere.normalize_points()
    write_ply(
        'textured_sphere.ply',
        viz_icosphere.get_vertices().transpose(0, 1).numpy(),
        rgb=rgb_vertices.cpu().numpy(),
        faces=viz_icosphere.get_all_face_vertex_indices().numpy(),
        text=False)

# ----------------------------------------------
# Compute SIFT descriptors for each patch
# ----------------------------------------------
tangent_image_kp, tangent_image_desc = sift_tangent_images(
    tex_image, corners, img.shape[-2:], crop_degree=0)
print('Num Tangent Image Keypoints:', tangent_image_kp.shape[0])

if save_ply:
    # Write tangent image keypoints to their 3D position on the sphere
    kp_3d = convert_spherical_to_3d(
        convert_image_to_spherical(tangent_image_kp[:, :2], img.shape[-2:]))

    # Write 3D keypoints
    write_ply('kp_tangent_images.ply', kp_3d.view(-1, 3).numpy().T)

# ---------------------------------------------------
# Compute SIFT descriptors on equirectangular image
# ---------------------------------------------------
erp_kp_details = sift_equirectangular(img, crop_degree=0)
erp_kp = erp_kp_details[0]
erp_desc = erp_kp_details[1]
print('Num Equirect. Keypoints:', erp_kp.shape[0])
if save_ply:
    kp_3d = convert_spherical_to_3d(
        convert_image_to_spherical(erp_kp[:, :2], img.shape[-2:]))
    write_ply('kp_equirectangular.ply', kp_3d.view(-1, 3).numpy().T)

# ----------------------------------------------
# Plot descriptors
# ----------------------------------------------

# Convert image back to numpy format
img = img.permute(1, 2, 0).byte().numpy()

# Create a figure and axis handle
fig, ax = plt.subplots(1, 1)

# Set up the plot
ax.set_aspect(1, adjustable='box')
ax.imshow(img)
ax.get_xaxis().set_ticks([])
ax.get_yaxis().set_ticks([])
ax.plot(erp_kp[:, 0], erp_kp[:, 1], 'r.', markersize=3.0)
ax.plot(tangent_image_kp[:, 0], tangent_image_kp[:, 1], 'b.', markersize=3.0)
plt.show()


experiments/sift/view_overlap.py
import torch
import torch.nn.functional as F
from mapped_convolution.transforms import *
from mapped_convolution.nn import *
from mapped_convolution.util import *

import os
import os.path as osp
import json
from skimage import io
import math
sin = math.sin
cos = math.cos


def load_rgb(root_path, basename):
    """Loads RGB"""
    img = torch.from_numpy(
        io.imread(osp.join(root_path, 'rgb',
                           basename + '_rgb.png'))[..., :3]).permute(2, 0, 1)
    mask = ~(img == 0).all(0)
    return img, mask.float()


def load_depth(root_path, basename):
    """Loads and pre-processing depth image"""
    depth = io.imread(osp.join(root_path, 'depth', basename + '_depth.png'))
    # missing values are encoded as 2^16 - 1
    missing_mask = (depth == 2**16 - 1)
    # depth 0..128m is stretched to full uint16 range (1/512 m step)
    depth = depth.astype(np.float32) / 512.0
    # clip to a pre-defined range

    depth = np.clip(depth, 0.0, 10.0)
    # zero out missing values
    depth[missing_mask] = 0.0

    return torch.from_numpy(depth)


def load_pose(root_path, basename):
    """Loads camera pose"""
    with open(osp.join(root_path, 'pose', basename + '_pose.json'), 'r') as f:
        data = json.load(f)
    return torch.tensor(data['camera_rt_matrix'])


def get_bearing_vectors(img_shape):
    # Compute bearing vectors per pixel
    lat, lon, _, _ = equirectangular_meshgrid(img_shape)
    xyz = convert_spherical_to_3d(torch.stack((lon, lat), -1))
    bearing = F.normalize(xyz, dim=-1)  # H x W x 3
    bearing = bearing.permute(2, 0, 1).contiguous()  # 3 x H x W
    return bearing


def get_in_bound_keypoints(right_img_rgb, left_img_depth, right_img_depth,
                           left_img_pose, right_img_pose, right_keypoints):
    """
    Transforms the keypoints from the right image into the left image, returning only the ones that are in bounds

    Returns K x 2 in-bounds points
    """

    # Compute bearing vectors per pixel
    bearing = get_bearing_vectors(right_img_rgb.shape[-2:])

    # Fast way to quickly get depth and bearing for keypoints only
    # Resulting shape: M, M x 3
    nn_interp_layer = Unresample('nearest')
    bisph_interp_layer = Unresample('bispherical')
    right_keypoint_depth = nn_interp_layer(
        right_img_depth.unsqueeze(0).unsqueeze(0),
        right_keypoints.unsqueeze(0)).squeeze()
    right_keypoint_bearing = bisph_interp_layer(
        bearing.unsqueeze(0), right_keypoints.unsqueeze(0)).squeeze().permute(
            1, 0)

    # Right keypoints converted to 3D points
    right_pts_3d = right_keypoint_bearing * right_keypoint_depth.view(-1, 1)
    # write_ply('right_kp.ply', right_pts_3d.permute(1, 0).numpy())

    # Transform right keypoints into left camera space
    proj_pts_3d = canonical_to_camera_transform(
        camera_to_canonical_transform(right_pts_3d, right_img_pose),
        left_img_pose)

    # Get the depth of the projected points
    proj_pts_depth = proj_pts_3d.norm(dim=-1)

    # Get the bearing vectors of the projected points
    proj_pts_bearing = F.normalize(proj_pts_3d, dim=-1)

    # Convert to 2D coordinates
    proj_in_left_img = convert_spherical_to_image(
        convert_3d_to_spherical(proj_pts_bearing), left_img_depth.shape)

    # Get the left image depth at those coordinates
    proj_in_left_depth = nn_interp_layer(
        left_img_depth.unsqueeze(0).unsqueeze(0),
        proj_in_left_img.unsqueeze(0)).squeeze()

    # Valid projected points are ones that have a lower depth than the left image depth (i.e. are visible)
    mask = proj_pts_depth < proj_in_left_depth
    proj_in_bounds = proj_in_left_img[mask]

    # Returns the K x 2 locations of the in-bounds points in the left image
    # Also returns the indicies of the right keypoints matrix that are inbounds
    return proj_in_bounds, mask.nonzero().squeeze()


# -------------------------------------------------
# -------------------------------------------------
# -------------------------------------------------

data_root = '/net/vision29/data/c/2D-3D-Semantics/area_1/pano'
left_img_basename = 'camera_af500027f89e4befa60da622257156ac_WC_1_frame_equirectangular_domain'
right_img_basename = 'camera_36dd48fe958d4699ae3de7eaf4b11eb3_WC_1_frame_equirectangular_domain'
base_order = -1
sample_order = 10
scale_factor = 1
img_shape = (2048 // scale_factor, 4096 // scale_factor)


def bilinear_rescale(img, scale):
    return F.interpolate(
        img.float().unsqueeze(0),
        scale_factor=scale,
        mode='bilinear',
        align_corners=False).byte().squeeze(0)


def nn_rescale(img, scale):
    return F.interpolate(
        img.unsqueeze(0).unsqueeze(0), scale_factor=scale,
        mode='nearest').squeeze()


def camera_to_canonical_transform(pts, pose):
    """
    pts: N x 3
    pose 3 x 4
    """
    return (pts - pose[:, 3]) @ pose[:3, :3]


def canonical_to_camera_transform(pts, pose):
    """
    pts: N x 3
    pose 3 x 4
    """
    return pts @ pose[:3, :3].T + pose[:, 3]


# Load data
left_img_rgb, left_img_mask = load_rgb(data_root,
                                       left_img_basename)  # 3 x H x W, H x W
right_img_rgb, right_img_mask = load_rgb(data_root,
                                         right_img_basename)  # 3 x H x W, H x W
# When loading depth, mask out the boundaries of the image
left_img_depth = left_img_mask * load_depth(data_root,
                                            left_img_basename)  # H x W
right_img_depth = right_img_mask * load_depth(data_root,
                                              right_img_basename)  # H x W

left_img_pose = load_pose(data_root, left_img_basename)  # 3 x 4
right_img_pose = load_pose(data_root, right_img_basename)  # 3 x 4

left_img_rgb = bilinear_rescale(left_img_rgb, 1 / scale_factor)
right_img_rgb = bilinear_rescale(right_img_rgb, 1 / scale_factor)
left_img_depth = nn_rescale(left_img_depth, 1 / scale_factor)
right_img_depth = nn_rescale(right_img_depth, 1 / scale_factor)
left_img_mask = nn_rescale(left_img_mask, 1 / scale_factor)
right_img_mask = nn_rescale(right_img_mask, 1 / scale_factor)


def compute_fov_overlap_mask(left_img_depth,
                             right_img_depth,
                             left_img_pose,
                             right_img_pose,
                             eps=0.05):

    bearing = get_bearing_vectors(right_img_depth.shape)
    right_pts = right_img_depth.unsqueeze(0) * bearing
    right_pts = right_pts.view(3, -1)
    right_pts = right_pts.permute(1, 0)
    t_right_pts = canonical_to_camera_transform(
        camera_to_canonical_transform(right_pts, right_img_pose), left_img_pose)

    # Compute the number of pixels in the right image that are visible in the left image
    # Depth of the projected points
    t_right_depth = t_right_pts.norm(dim=-1)

    # Bearing vectors of the projected points
    t_right_bearing = F.normalize(t_right_pts, dim=-1)

    # Convert to 2D coordinates
    t_right_in_left_img = convert_spherical_to_image(
        convert_3d_to_spherical(t_right_bearing), left_img_depth.shape)

    # Get the left image depth at those coordinates
    nn_interp_layer = Unresample('nearest')
    t_right_in_left_depth = nn_interp_layer(
        left_img_depth.unsqueeze(0).unsqueeze(0),
        t_right_in_left_img.unsqueeze(0)).squeeze()

    # Valid projected points are ones that have a lower depth than the left image depth (i.e. are visible)
    valid_right_in_left_mask = t_right_depth <= t_right_in_left_depth + eps

    return right_pts, valid_right_in_left_mask


def compute_fov_overlap(left_img_depth,
                        right_img_depth,
                        left_img_pose,
                        right_img_pose,
                        eps=0.05,
                        save_overlap_img=False):
    right_pts, valid_right_in_left_mask = compute_fov_overlap_mask(
        left_img_depth, right_img_depth, left_img_pose, right_img_pose)
    left_pts, valid_left_in_right_mask = compute_fov_overlap_mask(
        right_img_depth, left_img_depth, right_img_pose, left_img_pose)

    right_depth_mask = right_img_depth.view(-1) >= 1e-6
    left_depth_mask = left_img_depth.view(-1) >= 1e-6
    valid_pt_mask = right_depth_mask & left_depth_mask & right_img_mask.view(
        -1).bool() & left_img_mask.view(-1).bool()
    mask = valid_right_in_left_mask & valid_left_in_right_mask & valid_pt_mask
    num_valid_right = right_pts[right_depth_mask].shape[0]
    num_valid_left = left_pts[left_depth_mask].shape[0]

    right_fov_percent = (valid_right_in_left_mask
                         & valid_pt_mask).sum().float() / num_valid_right
    left_fov_percent = (valid_left_in_right_mask
                        & valid_pt_mask).sum().float() / num_valid_right

    valid_left_img_mask = (
        valid_left_in_right_mask * valid_pt_mask).view(*left_img_depth.shape)
    valid_right_img_mask = (
        valid_right_in_left_mask * valid_pt_mask).view(*left_img_depth.shape)

    if save_overlap_img:
        left_img_overlap = left_img_rgb.clone()
        left_img_overlap[0, valid_left_img_mask] = 200
        io.imsave('left_img_overlap.png',
                  left_img_overlap.permute(1, 2, 0).numpy())
        right_img_overlap = right_img_rgb.clone()
        right_img_overlap[1, valid_right_img_mask] = 200
        io.imsave('right_img_overlap.png',
                  right_img_overlap.permute(1, 2, 0).numpy())

    return (right_fov_percent + left_fov_percent) / 2


avg_fov_overlap = compute_fov_overlap(
    left_img_depth,
    right_img_depth,
    left_img_pose,
    right_img_pose,
    save_overlap_img=True)
print('Average FOV Overlap:', avg_fov_overlap)

# left_overlap = compute_fov_overlap_mask(right_img_depth, left_img_depth,
#                                           right_img_pose, left_img_pose)
# print('Right Overlap Metric:', right_overlap)
# print('Left Overlap Metric:', left_overlap)

# # # ---------------------------------

# # Compute the keypoints on the left and right images
# # Keypoints: M x 2
# # -----------------------------
# # IF PATCH:
# resample_to_uv_layer, corners = get_tangent_plane_info(base_order, sample_order,
#                                                        left_img_rgb.shape[-2:])
# right_tex_image = resample_to_uv_layer(
#     right_img_rgb.float().unsqueeze(0)).squeeze(0).byte()

# right_keypoints = extract_sift_feats_patch(right_tex_image,
#                                            corners,
#                                            image_shape=right_img_rgb.shape[-2:],
#                                            crop_degree=30)[:, :2].contiguous()

# left_tex_image = resample_to_uv_layer(
#     left_img_rgb.float().unsqueeze(0)).squeeze(0).byte()
# left_keypoints = extract_sift_feats_patch(left_tex_image,
#                                           corners,
#                                           image_shape=left_img_rgb.shape[-2:],
#                                           crop_degree=30)[:, :2].contiguous()

# in_bound_right_keypoints, right_matches = get_in_bound_keypoints(
#     right_img_rgb, left_img_depth, right_img_depth, left_img_pose,
#     right_img_pose, right_keypoints)
# print('Patch Level {}'.format(base_order))
# print('  Number of right keypoints', right_keypoints.shape[0])
# print('  Number of right keypoints in-bounds in left image:',
#       in_bound_right_keypoints.shape[0])

# # -----------------------------
# # -----------------------------
# # IF ERP:
# right_keypoints = extract_sift_feats_erp(right_img_rgb,
#                                          crop_degree=30)[:, :2].contiguous()

# left_keypoints = extract_sift_feats_erp(left_img_rgb,
#                                         crop_degree=30)[:, :2].contiguous()

# in_bound_right_keypoints, right_matches = get_in_bound_keypoints(
#     right_img_rgb, left_img_depth, right_img_depth, left_img_pose,
#     right_img_pose, right_keypoints)
# print('Equirectangular')
# print('  Number of right keypoints', right_keypoints.shape[0])
# print('  Number of right keypoints in-bounds in left image:',
#       in_bound_right_keypoints.shape[0])

# BELOW IS ONLY USEFUL FOR PRINTING A VISUALIZATION IMAGE FOR SANITY CHECKING
# orig_keypoints = [cv2.KeyPoint(k[0], k[1], 3.0, 0) for k in right_keypoints]
# proj_keypoints = [
#     cv2.KeyPoint(k[0], k[1], 3.0, 0) for k in in_bound_right_keypoints
# ]
# left_matches = torch.arange(in_bound_right_keypoints.shape[0])
# matches = torch.stack((left_matches, right_matches), -1)
# matches = [cv2.DMatch(m[0], m[1], 0) for m in matches]

# out_img = cv2.drawMatches(left_img_rgb.permute(1, 2, 0).numpy(),
#                           proj_keypoints,
#                           right_img_rgb.permute(1, 2, 0).numpy(),
#                           orig_keypoints,
#                           matches,
#                           None,
#                           matchColor=(255, 0, 255))
# io.imsave('reproject-sift-viz.png', out_img)
